Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,475 @@
+        set_frame_style(m_previous_frame_style.value());
+        m_previous_frame_style = {};
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
+    }
+
+    if (view_state() == ViewState::Expanded && !m_previous_frame_style.has_value() && height() == (content_margins().top() + content_margins().bottom() + m_controls_spacer->height())) {
+        m_controls_spacer->set_visible(false);
+        m_previous_frame_style = frame_style();
+        set_frame_style(Gfx::FrameStyle::NoFrame);
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
+    }
+}
+
+void DynamicWidgetContainer::child_event(Core::ChildEvent& event)
+{
+    if (event.type() == Event::ChildAdded && event.child() && is<GUI::DynamicWidgetContainer>(*event.child()))
+        static_cast<DynamicWidgetContainer&>(*event.child()).set_parent_container(*this);
+
+    GUI::Frame::child_event(event);
+}
+
+void DynamicWidgetContainer::set_parent_container(RefPtr<GUI::DynamicWidgetContainer> container)
+{
+    m_parent_container = container;
+}
+
+bool DynamicWidgetContainer::check_has_move_target(Gfx::IntPoint relative_mouse_position, MoveTargetOp operation)
+{
+    bool matched = false;
+    for_each_child_container([&](auto& child) {
+        bool is_target = child.relative_rect().contains(relative_mouse_position);
+        matched |= is_target;
+        child.set_render_as_move_target(operation == MoveTargetOp::SetTarget ? is_target : false);
+
+        return IterationDecision::Continue;
+    });
+    return matched;
+}
+
+void DynamicWidgetContainer::set_render_as_move_target(bool is_target)
+{
+    if (m_render_as_move_target == is_target)
+        return;
+    m_render_as_move_target = is_target;
+    update();
+}
+
+void DynamicWidgetContainer::swap_widget_positions(RefPtr<Core::EventReceiver> source_widget, Gfx::IntPoint destination_positon)
+{
+    RefPtr<Core::EventReceiver> destination_widget;
+    for_each_child_container([&](auto& child) {
+        if (child.relative_rect().contains(destination_positon)) {
+            destination_widget = child;
+            return IterationDecision::Break;
+        }
+        return IterationDecision::Continue;
+    });
+
+    VERIFY(destination_widget);
+    VERIFY(source_widget);
+
+    if (source_widget == destination_widget)
+        return;
+
+    auto source_pos = children().find_first_index(*source_widget);
+    auto destination_pos = children().find_first_index(*destination_widget);
+
+    VERIFY(source_pos.has_value());
+    VERIFY(destination_pos.has_value());
+
+    swap(children()[source_pos.value()], children()[destination_pos.value()]);
+
+    // FIXME: Find a better solution to instantly display the new widget order.
+    //        invalidate_layout is not working :/
+    auto childs = child_widgets();
+    for (RefPtr<GUI::Widget> widget : childs) {
+        widget->remove_from_parent();
+        (void)try_add_child(*widget);
+    }
```

Comment:
I'm pretty sure changes caused by modifying the child list directly are not handled correctly. Can you use the normal APIs instead?

Reply:
I was trying to improve that by using `remove_child` and `insert_child_before` but that didn't worked out. Somehow it was never working as expected and introduced some weird looking spaces between the moved widgets. For to moment i would like to leave the not perfect but working solution :-/